### PR TITLE
feat(openapi): cookie parameters

### PR DESCRIPTION
**Pull request by RafalFilipek**
### Motivation

When building a server with endpoints that must explicitly declare which cookies they require — and receive typed access to those values as part of the request input — there was no way to express this through oRPC's OpenAPI integration. The generated spec had no `in: cookie` parameter entries, and the decoded input object had no `cookies` key.

While oRPC provides cookie helpers (`getCookie`, `setCookie`, `deleteCookie`) for manual cookie access within handlers, this PR addresses a different need: treating cookies as **first-class, typed input parameters** — declared in the schema, validated at the boundary, and visible in the generated OpenAPI spec — in the same way `query`, `headers`, and `params` are handled today.

### What this does

Adds `cookies` as a supported key in `inputStructure: 'detailed'` mode:

**Spec generation** (`openapi-generator.ts`): the generator now recognises `cookies` as a fifth top-level key alongside `params`, `query`, `headers`, and `body`. Each property of the `cookies` schema is emitted as an `in: cookie` parameter object in the OpenAPI spec.

**Server-side decode** (`openapi-codec.ts`): the `decode()` method now exposes a lazy `cookies` getter on the detailed input object. It parses the incoming `Cookie` request header (using the same lazy-evaluation pattern already used for `query`) and returns a `Record<string, string>`. Uses `flattenHeader` from `@orpc/standard-server` for correct header normalisation.

### Usage

```ts
const getUser = os
  .route({ method: 'GET', path: '/me', inputStructure: 'detailed' })
  .input(z.object({
    cookies: z.object({
      session_id: z.string(),
      theme: z.string().optional(),
    }),
  }))
  .handler(({ input }) => {
    const { session_id, theme } = input.cookies // fully typed
    // ...
  })
```

Generated spec:

```
parameters:
  - name: session_id
    in: cookie
    required: true
    schema:
      type: string
  - name: theme
    in: cookie
    required: false
    schema:
      type: string
```

What is NOT included in this PR (initial implementation)

- inputStructure: 'compact' — cookies are not supported in compact mode; they only make sense in detailed mode where input fields are explicitly mapped to HTTP locations
- Client-side encoding (openapi-link-codec.ts) — the client does not yet encode input.cookies into outgoing Cookie headers
- Set-Cookie in output structure — response cookies are out of scope for this PR

I'd love to hear your feedback on this approach. If there are aspects of the design you'd like to see changed, or if you have suggestions on how to best continue this work (e.g. compact mode support, client-side encoding), I'm happy to follow your guidance.

Thank you,

Rafał

## Summary by CodeRabbit

* **Documentation**
  * Added a design spec and implementation plan for OpenAPI cookie-parameter support with usage examples and testing guidance.

* **New Features**
  * Enabled cookie parameters for detailed input structures and exposed parsed request cookies alongside params/query/headers.

* **Tests**
  * Added tests covering cookie parameter generation, parsing from request headers, empty/missing behavior, decoding edge cases, and validation rules.

### CODE REVIEW

File: packages/openapi/src/adapters/standard/openapi-codec.ts [hunk 1]
```diff
@@ -20,6 +21,19 @@ export interface StandardOpenAPICodecOptions {
...
+    return {}
+  return Object.fromEntries(
+    cookieHeader.split(';').filter(Boolean).flatMap((pair) => {
+      const idx = pair.indexOf('=')
+      if (idx === -1)
+        return []
+      return [[pair.slice(0, idx).trim(), pair.slice(idx + 1).trim()]]
+    }),
+  )
+}
```
[the remark this answers is not in the document]
Comment (RafalFilipek) [reply packages/openapi/src/adapters/standard/openapi-codec.ts]: @copilot IMO we should use `cookie` package to align it with `packages/server/src/helpers/cookie.ts`

File: packages/openapi/src/adapters/standard/openapi-codec.ts [hunk 2]
```diff
@@ -67,6 +69,14 @@ export class StandardOpenAPICodec implements StandardCodec {
         Object.defineProperty(this, 'query', { value, writable: true })
       },
       headers: request.headers,
+      get cookies() {
+        const value = parseCookie(flattenHeader(request.headers.cookie) ?? '')
```
[the remark this answers is not in the document]
Comment (RafalFilipek) [reply packages/openapi/src/adapters/standard/openapi-codec.ts:73]: Not a valid concern in this codebase. Header keys are guaranteed lowercase before reaching the codec by every supported adapter:

- fetch — Headers.forEach() yields lowercase keys per the WHATWG spec
- node — Node.js http.IncomingMessage.headers is lowercased by the HTTP/1.1 parser
- fastify — passes through req.raw.headers, same Node.js guarantee
- aws-lambda — AWS API Gateway V2 delivers headers lowercased at the runtime level
- 
This is also consistent with how the rest of the codebase accesses StandardHeaders — every lookup ('content-type', 'content-length', 'last-event-id', etc.) uses lowercase string literals. Treating cookie differently here would be inconsistent and address a gap that doesn't exist in practice for any supported adapter.

File: packages/openapi/src/adapters/standard/openapi-codec.ts [hunk 3]
```diff
@@ -67,6 +69,14 @@ export class StandardOpenAPICodec implements StandardCodec {
...
+        Object.defineProperty(this, 'cookies', { value, writable: true })
+        return value
+      },
```
[the remark this answers is not in the document]
Comment (RafalFilipek) [reply packages/openapi/src/adapters/standard/openapi-codec.ts:76]: Not a valid concern. `cookie@1.1.1` has a built-in `try/catch` in its internal decode function:

```
function decode(str) {
  if (str.indexOf("%") === -1) return str;
  try {
    return decodeURIComponent(str);
  } catch (e) {
    return str; // returns raw string on malformed encoding, never throws
  }
}
```


`cookie.parse` never throws on malformed percent-encoding — it falls back to the raw string.

**Comment by dinwwwh** [comment]
I think this is good, but maybe I should save it for v2. What do you think? @RafalFilipek

**Comment by dinwwwh** [comment]
My only concern with this feature is that the client can't programmatically set arbitrary `Cookie` headers in a browser - so if `cookies` shows up as part of `input`, it implies the client controls/sends that value, which isn't true and could mislead consumers of the generated spec.

If the goal is just to make cookie expectations visible in the generated OpenAPI spec (while server code still reads cookies manually via `getCookie`/`getCookie` helpers), I think that's actually simpler to achieve in v2, no input/decode changes needed, just a special meta key (e.g. `.route({ ... })` meta) that tells the generator to emit `in: cookie` parameters for documentation purposes only.

Happy to help explore that direction if useful.

**Comment by RafalFilipek** [comment]
Thanks for the thoughtful feedback, @dinwwwh — happy to wait for v2 if that's the better home for this.

To give you more context on our use case: we're building a server-side API where our security team requires that every authenticated endpoint explicitly declares the cookies it depends on — both for validation purposes and so that external systems (e.g. API gateways, security auditing tools) can inspect the contract and verify compliance. The `cookies` key in `input` is intentional: it expresses "this endpoint requires these cookies to be present, and here is their schema."

We're aware of the `getCookie`/`setCookie` helpers and do use them elsewhere. The difference here is that we need cookies to be part of the validated, typed contract — not just read ad-hoc in handler code. For this specific case we want cookies treated as first-class input parameters, on the same footing as `query` or `headers`.

Regarding the browser concern — you're right, and we're fully aware of that limitation. Browsers manage the `Cookie` header automatically; a client-side SDK cannot set it programmatically for cross-origin requests. This does require additional orchestration on the client side, and in other projects where we face similar situations we handle it through TypeScript-level tricks (e.g. narrowing or omitting the `cookies` field from the client-facing input type) so the generated client stays ergonomic without exposing something it can't actually control. It's not ideal, but it works.

That said, thinking about it more — you might be right that the cleanest separation is to keep cookie *reading* as it is today (via middleware and the `getCookie` helper, which gives a perfectly decent developer experience) and focus the framework-level feature purely on enriching the generated OpenAPI spec: a lightweight, declarative way to say "this endpoint expects these cookies" for documentation and external tooling, without changing how cookies are actually read at runtime.

Ultimately, though, we're completely open to whatever direction you think is best. Our main goal is that the solution is easy to reason about for new users and keeps the framework's mental model as clear as possible — even if that means our current workaround stays a workaround for a bit longer. If you have a direction in mind for v2, I'd love to hear it and contribute where we can :)

Thanks again for your time and for maintaining such a well-designed library.

**Comment by dinwwwh** [comment]
@RafalFilipek Thanks for the PR and the suggestion. I've considered this feature many times, but I've decided not to support it in either v1 or v2. Instead, I'll add helper utilities to make adding cookies to the OpenAPI spec easier.
